src/relax/ir/dataflow_matcher.cc

```diff
@@ -541,15 +541,21 @@ struct RNode {
  * \brief This method try to match a real node and a pattern node along with its neighbors.
  */
 static bool try_match(PNode* p, RNode* r, DFPatternMatcher* m,
-                      const std::map<const VarNode*, std::set<const VarNode*>>& def2use,
+                      const std::map<const VarNode*, std::vector<const VarNode*>>& def2use,
                       const std::map<const VarNode*, std::vector<const VarNode*>>& use2def) {
-  if (nullptr != p->matched && p->matched == r->ptr) return true;  // matched before.
+  if (p->matched != nullptr && p->matched == r->ptr) return true;  // matched before.
   if (!m->Match(GetRef<DFPattern>(p->ptr), GetRef<Var>(r->ptr))) return false;
 
   std::stack<std::pair<PNode*, RNode*>> undo_stack{};
 
   const auto commit = [&undo_stack](PNode* p, RNode* r) {
     // match with each other.
+    // TODO(ganler, masahi): Why commit on the same p-r pair happens more than once?
```

Comment:
I guess in either way the eventual results are the same. But yeah doing a pre-check could be faster (avoid the overhead of `undo_stack.emplace(p, r)`).

Reply:
Here my concern is not about performance. If the same pair is committed more than once, I think there is something odd about the matching algorithm.

Later I'll try to improve the matching algorithm implementation. In particular, I want to remove `try_match` loop on parents, https://github.com/apache/tvm/blob/278bc9bbdf7d9b12c44bb78b71f91e4872b14353/src/relax/ir/dataflow_matcher.cc#L577-L607. This bidirectional matching has been a source of confusion to me - Since the constraint graph is assumed to be a DAG, I think we can start matching from the root nodes in a topo-sorted order, and matching should be able to proceed purely in one direction.

Reply:
I see. Because the constraint or relation between nodes may or may not be single-way so in the beginning I made it bidirectional such that the pattern can be matched as long as you can let any node of the matched subgraph be start hint and get the pattern matched eventually. Specifying certain matching order definitely makes the code logic and debugging easier but I am afraid it also cuts the flexibility in some way. Not sure if it is worth to have some flexibility or if we can keep them both.

Reply:
I'm also going to look at how similar projects like MLIR, OpenVINO etc implements general graph pattern matching.

Reply:
Sorry, I need some time to illustrate but my flight is taking off now. I will get you back probably after 8 hours. 🥲

Reply:
ok :) I'll share my thought in the meantime. 

Looking at an example in https://github.com/apache/tvm/blob/unity/tests/python/relax/test_dataflow_pattern.py#L540-L545, I cannot imagine how one would use `start_hint` in practice. Here, you wrote the input mod by hand, so you know `dfb.bindings[0].var` is associated to the "left" branch of the CBRx2. But in general, we don't have such information, especially in e2e scenarios for real world models. 

Even if there was a good use case for it, I claim that it doesn't justify making API and the implementation more complicated. If one has such advanced knowledge of model structure and variables, they may as well have a different way to match & extract such subgraph.

Reply:
Thanks for bringing it up. When designing the usages it was my bad that I did not know enough use cases and I made it "more capable" by compromising simplicity. masa you have more professional experience with how patterns look in practice so I strongly agree that we should cut off the implementation/complexity for such long-tail use cases.

Reply:
BTW, also to explain why I was using the code style of `CONSTANT logic_op VAR`. This is just a practice to avoid writing `VAR = CONSTANT` which causes silent logic errors. But yeah it is nice to normalize the code style.